whd/dispatcher.py

```diff
@@ -213,6 +219,130 @@ def _set_pr_labels(self, pr_event, resources):
                     f" in '{repository_path}' found. Ignoring, since they are not an org member'."
                 )
 
+    def _resource_versions_and_jobs(self, concourse_api, matching_resources, pr_event):
+        resource = [r for r in matching_resources if r.type == "pull-request"]
+        if len(resource) != 1:
+            logger.warning(f"none or more than on resource found for PR {pr_event.number()}")
+            return
+
+        resource = resource[0]
+        resource_versions = concourse_api.resource_versions(
+            pipeline_name=resource.pipeline.name,
+            resource_name=resource.name
+        )
+        if not resource_versions:
+            logger.warning(
+                f"no resource versions found for pipeline {resource.pipeline.name} "
+                f"and resource {resource.name}"
+            )
+            return
+
+        pr_resource_versions = [
+            rv for rv in resource_versions if rv.version()['pr'] == str(pr_event.number())
+        ]
+        if not pr_resource_versions:
+            logger.warning(
+                f"no resource versions found for pipeline {resource.pipeline.name} "
+                f"resource {resource.name} and PR {pr_event.number()}"
+            )
+            return
+
+        jobs_with_triggering_resource = [
+            job for job in resource.pipeline.jobs if job.is_triggered_by_resource(resource.name)
+        ]
+
+        # check if there is a pr resource version which was not build, if so return
+        # it with the corresponding job
+        for job in jobs_with_triggering_resource:
+            builds = concourse_api.job_builds(job.pipeline.name, job.name)
+            for resource_version in self._resource_version_ids_not_built(
+                resource_versions=pr_resource_versions,
+                builds=builds
+            ):
+                yield (job, resource, resource_version)
+
+    def _ensure_pull_requests_jobs_are_run(self, concourse_api, matching_resources, pr_event):
+        '''
+        check all PR resource versions for which no build exists. If a resource version has
+        no build, pin this resource version, trigger the corresponding job and unpin the resource
+        '''
+        for job, resource, resource_version in self._resource_versions_and_jobs(
+            concourse_api, matching_resources, pr_event
+        ):
+            logger.info(
+                f"Pin resource version {resource_version.version()} "
+                f"and trigger build for job {job.name}"
+            )
+            self._trigger_build_with_pinned_resource_version(
+                job=job,
+                resource_name=resource.name,
+                resource_version_id=resource_version.id()
+            )
+
+    def _resource_version_ids_not_built(self, resource_versions, builds):
```

Comment:
could you please rather use a more generic term such as `triggered`, or `run`, or `executed`. not all jobs are builds. also, what the jobs do is completely irrelevant at this point. we care about them to be triggered.